### ISSUE TITLE
[inema] Can not use custom prices for products

### ISSUE THREAD

**Issue by novawish** [gsauthof/frank#3]
Hello everyone,

I have one problem with the library `inema` but I could not find its repository on Github so I created the issue here.
Currently if we call `checkout` (either PDF or PNG) it will calculate total price by 
`compute_total` which only uses the pre-defined data in json file (for example: `data/products-2020-01-01.json`).
I am having a case that if we have another price for some products, the API server will not allow us to checkout them unless we checkout with the exact prices we have in our account. And there is no possible way to pass our custom prices through function when using the library.

I would be glad to contribute a PR of the fix if it is possible, otherwise if you have any other solution please help me, thanks very much.

**Comment by gsauthof** [comment]
You mean that you have some customer specific prices for some products (e.g. discounts)?

Or are some prices in the bundled `products*json` just out of date?

**Comment by gsauthof** [comment]
One way to deal with the former situation is to overwrite the price for the affected product after import, e.g.:

```
[ins] In [1]: import inema

[ins] In [2]: inema.inema.marke_products['404']['cost_price']
Out[2]: '1.67'

[ins] In [3]: inema.inema.marke_products['404']['cost_price'] = '1.66'

[ins] In [4]: inema.inema.marke_products['404']['cost_price']
Out[4]: '1.66'
```

Would that be sufficient for your use case?

If we are talking about outdated price information in the package please let me know the details and I'll update the inema package.

**Comment by novawish** [comment]
I'm meaning that I have some customer specific prices.
Yes your solution works, although it is not very efficient because every time I call the library i have to overwrite the prices again. If there is no other way I guess I will stick with it then.
Thanks for you answer.

**Comment by gsauthof** [comment]
Do you have some way how to deal with this in mind?

What currently also is possibly is to drop a customer specific `products*json` into the package install directory. Just make sure that is has the newest date suffix. Of course, this doesn't help if your software has to deal with different customers (that have different price discounts) on the same machine.

One could also extend inema such that a `products*json` file in - say - `$HOME/.config/inema/` has priority over the one present in the package directory. Which would allow for user specific product configurations. Again, this wouldn't help if the software runs under one technical user for different customers that have different prices.

Yet another way would be to introduce another optional parameter to the `Internetmarke` constructor where the caller can specify her own products dictionary.

**Comment by novawish** [comment]
I have one idea that we could make the property `positions` as a list of tuple consists of 2 elements: `ptype` and `custom_price`.
The default value of parameter `custom_price` is **False** so that if the `custom_price` is not set, the function `compute_total` will use prices from `marke_products` dictionary.

So for example, the functions will be like this:
```python
def add_position(self, position, custom_price=False):
    _logger.info("Adding position to basket: %s", position)
    self.positions.append((position, custom_price))
```

```python
def compute_total(self):
    total = 0
    for prod, price in self.positions:
        if price:
            total += price
        else:
            total += self.get_product_price_by_id(prod.productCode)
    return total
```
And of course we have to modify the `checkout` functions to adapt the new structure of `positions`

**Comment by novawish** [comment]
Hi @gsauthof, did you have any feedback on this?

**Comment by novawish** [comment]
Hallo @gsauthof, can we move on with this?
I am not sure if you are aware but since today (1st of July 2020) the product prices are changed due to tax reduction in Germany.
The current implementation gives a very limited flexibility in these circumstances.

Of course I am always willing to contribute a Pull Request if possible.

**Comment by gsauthof** [comment]
I was busy with some other projects.

I think modifying `compute_total()` in that way could break existing code that uses inema.

What I had in mind in something like this:

```diff
--- a/inema/inema.py
+++ b/inema/inema.py
@@ -25,16 +25,16 @@ _logger = logging.getLogger(__name__)
 # This allows to ship multiple products.json files for announced future price/product
 # changes.
 def load_products():
-    global marke_products
     ps = [ e for e in resource_listdir(__name__, 'data')
             if e.startswith('products-') and e.endswith('.json')
             and e <= 'products-{}.json'.format(date.today().isoformat()) ]
     ps.sort()
     ps.insert(0, 'products.json')
     products_json = resource_stream(__name__, 'data/' + ps[-1]).read().decode("utf-8")
-    marke_products = json.loads(products_json)
+    products = json.loads(products_json)
+    return products
 
-load_products()
+default_products = load_products()
 
 formats_json = resource_stream(__name__, "data/formats.json").read().decode("utf-8")
 formats = json.loads(formats_json)
@@ -59,6 +59,8 @@ class ProductInformation(object):
 class Internetmarke(object):
     wsdl_url = 'https://internetmarke.deutschepost.de/OneClickForAppV3/OneClickForAppServiceV3?wsdl'
 
+
+
     # generate a 1C4A SOAP header
     def gen_1c4a_hdr(self, partner_id, key_phase, key):
         # Compute 1C4A request hash accordig to Section 4 of service description
@@ -94,16 +96,21 @@ class Internetmarke(object):
         return [p, t, k, s]
 
     def get_product_price_by_id(self, ext_prod_id):
-        price_str = marke_products[str(ext_prod_id)]['cost_price']
+        price_str = self.products[str(ext_prod_id)]['cost_price']
         return int(decimal.Decimal(price_str) * 100)
 
-    def __init__(self, partner_id, key, key_phase="1"):
+    def __init__(self, partner_id, key, key_phase="1", products=None):
         self.client = Client(self.wsdl_url)
         self.partner_id = partner_id
         self.key_phase = key_phase
         self.key = key
         self.soapheader = self.gen_1c4a_hdr(self.partner_id, self.key_phase, self.key)
         self.positions = []
+        if products is None:
+            self.products = default_products
+        else:
+            self.products = products
+
 
     def authenticate(self, username, password):
         s = self.client.service
```

You could then use it like this for customer specific prices:

```python
import inema

products_discount = inema.load_products()
# change it once
products_discount['10001'].cost_price = '0.99'

# ...

im = inema.InternetMarke(partner, key, products=products_discount)

# or even change it later

im = inema.InternetMarke(partner, key)
im.products['10001'].cost_price = '0.80'
```

What do you think, would that work for you?

**Comment by novawish** [comment]
Hi @gsauthof,
Yes i think it will work for me, however I would like to add an opinion on your solution:
- That approach will allow user to fully define their products dictionary. If they try to modify something in the dictionary, it might cause some troubles (product id is missing, `cost_price` is not provided, wrong data type of products...). Many possibilities can happen by this way.

Other than that, I am fine with your solution. Thanks very much for coming up with that.

Regards.

**Comment by gsauthof** [comment]
Ok, I've pushed it to master: https://git.sysmocom.de/python-inema/commit/?id=640baf1dbc6f09e02174d2617958fa2de5c6a0c3

I'll test it locally a few days before I cut the next PyPi release.

**Comment by gsauthof** [comment]
This change's included in today's 0.8.4 release.
